File: leet-code/404.左叶子之和.js
/*
 * @lc app=leetcode.cn id=404 lang=javascript
 *
 * [404] 左叶子之和
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var sumOfLeftLeaves = function (root) {
  if (!root) return 0
  // 叶子节点
  if (!root.left && !root.right) return 0
  // 左
  let leftNum = sumOfLeftLeaves(root.left)
  if (root.left && !root.left.left && !root.left.right) {
    // 如果是左叶子节点的父亲，左边num不再是叶子节点返回的0，而是左叶子节点的val
    leftNum = root.left.val
  }

  // 右，右节点正常递归查找是否存在左叶子节点
  const rightNum = sumOfLeftLeaves(root.right)

  // 中，累加左右节点的和
  return leftNum + rightNum
};
// @lc code=end


// @after-stub-for-debug-begin
module.exports = sumOfLeftLeaves;
// @after-stub-for-debug-end

File: leet-code-rs/src/solutions/404_左叶子之和.rs
/*
 * @lc app=leetcode.cn id=404 lang=rust
 *
 * [404] 左叶子之和
 */
#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}
// @lc code=start
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
//
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::cell::RefCell;
use std::rc::Rc;

#[allow(unused)]
impl Solution {
    pub fn sum_of_left_leaves(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
        fn get_left_leaves(node: &Option<Rc<RefCell<TreeNode>>>, sum: &mut i32, collected: bool) {
            if let Some(node) = node {
                let node = node.borrow();
                if node.left.is_none() && node.right.is_none() && collected {
                    // 左叶子
                    *sum += node.val;
                    return;
                }

                let left = &node.left;
                get_left_leaves(left, sum, true);

                let right = &node.right;
                get_left_leaves(right, sum, false);
            }
        }

        let mut sum = 0;
        get_left_leaves(&root, &mut sum, false);

        return sum;
    }
}
// @lc code=end

#[allow(unused)]
struct Solution;

#[cfg(test)]
mod test {
    #[allow(unused)]
    use super::*;

    #[test]
    fn test_sum_of_left_leaves() {
        //assert_eq!(Solution::sum_of_left_leaves(vec![]),[]);
        assert!(true)
    }
}
